src/pb_generator.cc

```diff
@@ -25,38 +25,110 @@
 // OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
 #include "pb_generator.h"
+#include <chrono>
+#include "pb_stub.h"
 
 #include <pybind11/embed.h>
 namespace py = pybind11;
 
 namespace triton { namespace backend { namespace python {
 
 ResponseGenerator::ResponseGenerator(
-    const std::vector<std::shared_ptr<InferResponse>>& responses)
-    : responses_(responses), index_(0)
+    const std::shared_ptr<InferResponse>& response)
+    : id_(response->Id()), is_finished_(false), is_cleared_(false)
 {
+  response_buffer_.push(response);
+}
+
+ResponseGenerator::~ResponseGenerator()
+{
+  if (!is_cleared_) {
+    Clear();
+  }
+  responses_.clear();
 }
 
 std::shared_ptr<InferResponse>
 ResponseGenerator::Next()
 {
-  if (index_ == responses_.size()) {
+  if (is_finished_) {
+    if (!is_cleared_) {
+      Clear();
+    }
     throw py::stop_iteration("Iteration is done for the responses.");
   }
 
-  return responses_[index_++];
+  std::shared_ptr<InferResponse> response;
+  {
+    {
+      std::unique_lock<std::mutex> lock{mu_};
+      while (response_buffer_.empty()) {
+        py::gil_scoped_release release;
+        cv_.wait(lock);
+      }
+      response = response_buffer_.front();
+      response_buffer_.pop();
+      is_finished_ = response->IsLastResponse();
+    }
+  }
+
+  if (is_finished_) {
+    Clear();
+    // Handle the case where the last response is empty.
+    if (response->IsEmptyResponse()) {
+      throw py::stop_iteration("Iteration is done for the responses.");
+    }
+  }
+
+  return response;
+}
+
+py::iterator
+ResponseGenerator::Iter()
+{
+  bool done = false;
+  while (!done) {
+    try {
+      responses_.push_back(Next());
+    }
+    catch (const py::stop_iteration& exception) {
+      done = true;
+    }
+  }
+
+  return py::make_iterator(responses_.begin(), responses_.end());
```

Comment:
I think we should return a new response generator here instead of fetching all the responses at once. This would allow the loop to make progress while the responses are being fetched from the backend.

Reply:
Returned self here.